pkgs/package_config/lib/src/package_config.dart

```diff
@@ -400,3 +400,85 @@ abstract class InvalidLanguageVersion implements LanguageVersion {
   @override
   String toString();
 }
+
+/// Relational operators for [LanguageVersion] that
+/// compare valid versions with [LanguageVersion.compareTo].
+///
+/// An [InvalidLanguageVersion] is not less or greater than other versions
+/// and is only equal to itself.
+// TODO(v3): Consider declaring on LanguageVersion class.
+extension LanguageVersionRelationalOperators on LanguageVersion {
+  /// Whether this language version is less than [other].
+  ///
+  /// An [InvalidLanguageVersion] is never less than or greater than
```

Comment:
Should `InvalidLanguageVersion` instead be an invalid input (throw `ArgumentError`? 

I cannot think of cases where I would like the program to just continue after I compare a valid version to an invalid version.

Reply:
I don't remember why the invalid language version was introduced. Probably to not break the code that called it on syntax errors, letting it handle it instead. (It was before null safety, so using `null` was even more error prone. Not sure that's still true today.)

You probably shouldn't keep using an invalid language version after recognizing that it's there. We could let the `<`'s all throw.
Then only `==` and `compareTo` works, because they're needed if you want to put it into a collection and/or sort it before checking for validity.

Reply:
Working on this PR made me wonder the value of the `InvalidLanguageVersion` in the first place. I'd much rather handle it being invalid when parsing, and as a user of the package, I'd prefer to know all `LanguageVersion` I get are valid.

As a result of this state, I think it's probably best to avoid throwing an error here as a user of the package might not recognize an `InvalidLanguageVersion` (potentially provided to them from elsewhere) is possible and could therefore throw an error here.

I'll defer to @lrhn here though. I'd also be interested to hear the original goals behind `InvalidLanguageVersion`. Thanks!

Reply:
Ah @lrhn commented before I did. Thanks!

> You probably shouldn't keep using an invalid language version after recognizing that it's there. We could let the <'s all throw.

I agree, but I'm not sure all users of the package would recognize that it's a possibility.

> I don't remember why the invalid language version was introduced. Probably to not break the code that called it on syntax errors, letting it handle it instead. (It was before null safety, so using null was even more error prone. Not sure that's still true today.)

I'd be in favor of removing it in the next major version as we have some improved ways to surface the error state now. We could also drop support for `.packages` then as well.

Reply:
(The invalid version object was introduced in https://github.com/dart-lang/package_config/commit/96ea9163328c16fc28e1518ed1daf7352eb8769d
without much discussion, other than letting the caller handle it instead throwing eagerly.
It was pretty certainly to allow parsing an entire `package_config.json` file and reporting multiple errors, rather than failing eagerly. I think it was @jensjoha who asked for it.)

Reply:
The CFE would like to continue compiling. Having something invalid is just another error.

Reply:
In that case I really think it makes sense for the comparison operators to throw.

If we want to be able to handle invalid versions the handling should be explicit.